libmcount/pyhook.c

```diff
@@ -0,0 +1,225 @@
+static PyAPI_FUNC(long) (*__PyLong_AsLong)(PyObject *);
+
+static PyObject *pName, *pModule, *pFuncEntry, *pFuncExit;
+
+#define INIT_PY_API_FUNC(func) \
+	do { \
+		__##func = dlsym(python_handle, #func); \
+		if (!__##func) { \
+			pr_err("dlsym for \"" #func "\" is failed!\n"); \
+			python_enabled = false; \
+			return -1; \
+		} \
+	} while (0)
+
+/**
+ * abs_dirname - parse given @pathname and set absolute dirname to @absdir
+ * @absdir: input buffer that will store absolute dirname
+ * @pathname: pathname string that can be either absolute or relative path
+ *
+ * This function parses the @pathname and sets absolute dirname to @absdir.
+ *
+ * Given @pathname sets @absdir as follows:
+ *
+ *    @pathname               | @absdir
+ *   -------------------------+---------------
+ *    mcount.py               | $PWD
+ *    tests/mcount.py         | $PWD/tests
+ *    ./tests/mcount.py       | $PWD/./tests
+ *    /root/uftrace/mcount.py | /root/uftrace
+ */
+static int abs_dirname(char *absdir, const char *pathname)
+{
+	assert(pathname);
+
+	char *path = strdup(pathname);
+	char *dir = ".";
+
+	if (strchr(path, '/'))
+		dir = dirname(path);
+
+	if (pathname[0] == '/') {
+		sprintf(absdir, "%s", dir);
+	}
+	else {
+		char cwd[BUFSIZ];
+		if (!getcwd(cwd, sizeof(cwd))) {
+			pr_err("getcwd error!\n");
+			return -1;
+		}
+		sprintf(absdir, "%s/%s", cwd, dir);
+	}
+	free(path);
+
+	return 0;
+}
+
+/* Import python module that is given by -p option */
+static int import_python_module(const char *py_arg_pathname)
+{
+	char py_sys_dir[BUFSIZ];
+	abs_dirname(py_sys_dir, py_arg_pathname);
+
+	/* Set path to import a python module. */
+	__PySys_SetPath(py_sys_dir);
+	pr_dbg("PySys_SetPath(\"%s\") is done!\n", py_sys_dir);
+
+	char *py_basename = basename((char*)py_arg_pathname);
+	pName = __PyString_FromString(py_basename);
+	pModule = __PyImport_Import(pName);
+	if (pModule == NULL) {
+		__PyErr_Print();
+		pr_warn("%s.py cannot be imported!\n", py_arg_pathname);
+		return -1;
+	}
+
+	return 0;
+}
+
+int python_init(const char *py_pathname)
+{
+	python_handle = dlopen(libpython, RTLD_LAZY);
+	if (!python_handle) {
+		pr_warn("%s cannot be loaded!\n", libpython);
+		return -1;
+	}
+
+	INIT_PY_API_FUNC(Py_Initialize);
+	INIT_PY_API_FUNC(PySys_SetPath);
+	INIT_PY_API_FUNC(PyErr_Print);
+	INIT_PY_API_FUNC(PyCallable_Check);
+	INIT_PY_API_FUNC(PyImport_Import);
+	INIT_PY_API_FUNC(PyTuple_New);
+	INIT_PY_API_FUNC(PyErr_Occurred);
+
+	INIT_PY_API_FUNC(PyTuple_SetItem);
+	INIT_PY_API_FUNC(PyObject_GetAttrString);
+	INIT_PY_API_FUNC(PyObject_CallObject);
+
+	INIT_PY_API_FUNC(PyString_FromString);
+	INIT_PY_API_FUNC(PyInt_FromLong);
+	INIT_PY_API_FUNC(PyString_AsString);
+	INIT_PY_API_FUNC(PyLong_AsLong);
+
+	__Py_Initialize();
+
+	/* import python module that is passed by -p option */
+	if (import_python_module(py_pathname) < 0)
+		return -1;
+
+	pFuncEntry = __PyObject_GetAttrString(pModule, "mcount_entry");
```

Comment:
It'd be great if the function name it calls is more generic - what about "uftrace_entry"?